### ISSUE TITLE
Support dictionary encoding with bitpacked indices

### ISSUE THREAD

**Issue by niyue** [lance-format/lance#2922]
Hi,

I would like to explore enhancing the Lance v2 format's dictionary encoding by integrating bitpacking to further reduce storage space. For instance, in a string array with 50 unique values, the default dictionary encoding would use an uint8 index. However, since 50 unique values can be represented with 6 bits (2^6 = 64 > 50), we could employ bitpacking to store dictionary indices in `6 * num_of_rows` bits rather than the current `8 * num_of_rows` bits.

I'm interested in working on this enhancement, but I want to ensure that it aligns with the project's evolution and avoids introducing code duplication. In reviewing the codebase, I noticed that while `BitpackedArrayEncoder` (located in `lance-encoding/src/encodings/physical/bitpack.rs`) exists, it is not currently in use. Additionally, the `BitpackingBufferEncoder`, which was introduced in PR #2333, was later removed for some reason.

I would appreciate insights on the following:

* Is such an enhancement desirable for the Lance v2 format?
* Do you have any suggestions on how to introduce bitpacking for dictionary encoding in a way that avoids code duplication and aligns with the current architecture?

Thanks.

**Comment by westonpace** [comment]
> Is such an enhancement desirable for the Lance v2 format?

Yes, this is something we want.

> BitpackedArrayEncoder (located in lance-encoding/src/encodings/physical/bitpack.rs) exists, it is not currently in use

It's not in use as it isn't yet stable and thoroughly tested and still under development somewhat (https://github.com/lancedb/lance/pull/2912) and (https://github.com/lancedb/lance/pull/2886).  However, I see no problem with connecting it up to be used as long as it's behind a 2.1 feature gate (this is how FSST is configured right now).  The spot where we pick which encoding to use is https://github.com/lancedb/lance/blob/5606b170a19f6ef3f8294a370395bf767655238d/rust/lance-encoding/src/encoder.rs#L211

> Additionally, the BitpackingBufferEncoder, which was introduced in PR https://github.com/lancedb/lance/pull/2333, was later removed for some reason.

It became BitpackedArrayEncoder.

> Do you have any suggestions on how to introduce bitpacking for dictionary encoding in a way that avoids code duplication and aligns with the current architecture?

I think we just need to choose it, as mentioned above.  We should use bitpacking in `choose_array_encoder` whenever:

 * The type is integral, or temporal
 * The version is 2.1 or higher
 * Bitpacking can actually provide an advantage

I believe there is logic to do it here: https://github.com/lancedb/lance/pull/2886/files  So you might try checking out that PR and testing if it works.  It should automatically start using bitpacking for dictioanary indices since we call `choose_array_encoder` to find the best encoding for the indices.  You'd just need to checkout that PR and then specify 2.1 as the file version in python:

```
from lance.file import LanceFileWriter

with LanceFileWriter("/some/path.lance", version="2.1") as writer:
  ...
```

**Comment by westonpace** [comment]
I don't think we have any tests specifically around dictionary + bitpacking (all the bitpacking unit tests focus on integer data).

Things that still remain to work on:

 * We don't have any (intentional) tests of dictionary + bitpacking.  I suppose, once bitpacking is enabled by #2886 then some of our existing dictionary tests will probably test this.  It wouldn't hurt to have a few python tests for dictionary encoding
 * Our use of dictionary encoding is still pretty conservative at the moment.  We only use it if there are fewer than 256 keys and only for the string type.  We need to support dictionary encoding for more types (large string, binary, large binary) and we should enable it with uint16 keys if it makes sense.

**Comment by broccoliSpicy** [comment]
> Do you have any suggestions on how to introduce bitpacking for dictionary encoding in a way that avoids code duplication and aligns with the current architecture?

there might be other issues come up when directly use `choose_array_encoer`'s `data_type` parameter to do encoding selection, reference: https://github.com/lancedb/lance/issues/2927
